### ISSUE TITLE
Record Real Supply and Demand Data

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#973]
**Is your feature request related to a problem? Please describe.**
- Currently, mars-sim uses primarily projected demand to come up with the demand value of a good.
- There may be huge disconnect between the real demand vs. projected demand.
- Demand, Supply, Value Point are currently dimenionless. 

**Describe the solution you desire to see**
- Record the real demand when calling the retrieve method. Record the real supply when calling the store method.
- Raise the average supply intelligently and automatically to meet the average demand of goods.

**Screenshots**
![image](https://github.com/mars-sim/mars-sim/assets/1168584/7b67d33f-7501-4616-b6c7-2ea784ee5bff)

**Specifications  (please complete the following information):**
 - pre 3.6.0 build 8421

**Additional context**
- We may codify supply, demand and consumption of resources as having the unit of `amount per sol` or `quantity per sol`.
- We may revise the existing *Demand* column to *Projected Demand* and add the *Real Demand* column. Likewise with the Supply.

**Comment by mokun** [comment]
@bevans2000 

Should I use `SolMetricDataLogger<Integer>` to record the demand of a resource ?

For Amount Resources and Parts, should I call the `increaseDataPoint` method inside MicroInventory ?

For Equipment, we will need to handle differently. May be within EquipmentInventory ?

**Comment by mokun** [comment]
Actually, I should call `increaseDataPoint` when resource being created in manufacturing, food production or resource processing. 

It's because it's no guaranteed that `storeAmountResource` really mean creation of resource. It could also be transferring resources.

Now that we mention that, may be we should create `transferAmountResource` and `transferItemResource` for those use cases if a resource is simply relocated from one ResourceHolder or ItemHolder to another, right ?

**Comment by bevans2000** [comment]
Only when taking and storing in a Settlement do we need to record the action. So the existing classes should not be changed. We probably should override the Settlement instance to record the amount so we don't have to change all the callers. 
Plus we don't need to record the full history so don't use DataLogger classes. We should just record the movements and then periodically count up the values to get new ratings and clear the history to start again.
Let's put this down for 3.7.0 as it needs some thought

**Comment by mokun** [comment]
> Plus we don't need to record the full history so don't use DataLogger classes. 

But I thought we've been using DataLogger to record the daily oxygen, CO2 and water usage and generation.

So you want ro simplify DataLogger or create an entirely new logger to do these type of daily statistics?

**Comment by bevans2000** [comment]
Thinking about this now it is quite easy. We only need to record the Supply & Demand for the Inventory of a Settlement so we can create a specialised subclass.
Then would we use the real values in the Goods Value calculations instead of the predicted values currently? The latter is very complicated.

**Comment by bevans2000** [comment]
Where do we want to use the real demand/supply? I can do it by building a Facade to the existing MicroInventory class that records the in & outs of the Inventory. But where to use them?

**Comment by bevans2000** [comment]
Also I think we should remove the ```Bin``` concept unless we are going to start using it. Having temporary holding code makes it difficult to read sometimes.

**Comment by mokun** [comment]
> Also I think we should remove the Bin concept unless we are going to start using it. 

Yes I've wanted to use the **Bin** concept for modeling those field-replaceable System or Equipment that are a part of a building.

I'll get back to that soon.

**Comment by mokun** [comment]
@bevans2000 @Urwumpe 

In order to understand more visually how each good will go up and down, we need more metrics.

I figure it's a good idea to show the flattened demand and the projected demand of each resource in the Trade Good tab. 

This way, we can compare goods that will the highest demand pressure in the foreseeable future. 

Having these two data points on each good will give us a glimpse on the future demand and the VP on each Good.

I'm almost ready to commit.

**Comment by mokun** [comment]
Also planning to add display the trade demand and the repair demand of each resource in the Goods tab in Monitor Tool.

**Comment by mokun** [comment]
For now, this is how it looks after adding these 4 values for each resource

Note that I highlighted the parts that a EVA Suit consists of in the pic below : 
![image](https://github.com/mars-sim/mars-sim/assets/1168584/d5f97e04-3c5d-47f6-812f-728ca936a985)
